python/cuml/cuml/sample/estimator.py

```diff
@@ -0,0 +1,137 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import numpy as np
+
+from cuml.internals.array import CumlArray
+from cuml.internals.global_settings import GlobalSettings
+from cuml.internals.mixins import FMajorInputTagMixin
+from cuml.internals.base import UniversalBase, DynamicDescriptor
+
+
+def io_fit(func):
+    def wrapper(self, *args, **kwargs):
+        # increase global counter to detect we are internal
+        GlobalSettings().increase_arc()
+
+        # check input type of first arg and fit estimator
+        self._set_output_type(args[0])
+        result = func(self, *args, **kwargs) 
+        self._is_fit = True
+
+        # decrease counter after exiting function
+        GlobalSettings().decrease_arc()
+
+        return result
+
+    return wrapper
+
+
+def io_predict_transform_array(func):
+    def wrapper(self, *args, **kwargs):
+        # increase global counter to detect we are internal
+        GlobalSettings().increase_arc()
+
+        result = func(self, *args, **kwargs) 
+
+        # decrease counter after exiting function
+        GlobalSettings().decrease_arc()
+
+        if GlobalSettings().is_internal:
+            return result
+
+        else:
+            # need to add logic to check globalsettings and mirror output_type
+            return result.to_output(self._input_type) 
+
+        return result
+
+    return wrapper
+
+
+class Estimator(UniversalBase,
+                FMajorInputTagMixin):
+    coef_ = DynamicDescriptor("coef_")
+    intercept_ = DynamicDescriptor("intercept_")
+
+    def __init__(self, 
+                 *, 
+                 awesome=True,
+                 output_type=None,
+                 handle=None,
+                 verbose=None):
+
+        super().__init__(handle=handle,
+                         verbose=verbose,
+                         output_type=output_type)
+
+        self.awesome = awesome
+        self._is_fit = False  # this goes in base
+
+    @io_fit
+    def fit(self,
+            X,
+            y,
+            convert_dtype=True):
+
+        input_X = CumlArray.from_input(
+            X,
+            order="C",
+            convert_dtype=convert_dtype,
+            target_dtype=np.float32,
+            check_dtype=[np.float32, np.float64],
+        )
+        self.n_features_in_ = input_X.n_cols
+        self.dtype = input_X.dtype
+
+        input_y = CumlArray.from_input(
+            y,
+            order="C",
+            convert_dtype=convert_dtype,
+            target_dtype=self.dtype,
+            check_dtype=[np.float32, np.float64],
+        )
+
+        self.coef_ = CumlArray.zeros(self.n_features_in_, 
+                                     dtype=self.dtype)
+
+        self.intercept_ = CumlArray.zeros(self.n_features_in_, 
+                                     dtype=self.dtype)
+
+        # do awesome C++ fitting here :) 
```

Comment:
One weird thing while playing with this a bit: when I add a `print(f"{self.coef_=}")` here I get the following:
```
Traceback (most recent call last):
  File "/home/coder/cuml/../ff.py", line 9, in <module>
    e.fit(X, y)
  File "/home/coder/.conda/envs/rapids/lib/python3.12/site-packages/cuml/internals/api_decorators.py", line 190, in wrapper
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/home/coder/.conda/envs/rapids/lib/python3.12/site-packages/cuml/sample/estimator.py", line 32, in wrapper
    result = func(self, *args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/coder/.conda/envs/rapids/lib/python3.12/site-packages/cuml/sample/estimator.py", line 115, in fit
    print(f"{self.coef_=}")
             ^^^^^^^^^^
  File "base.pyx", line 337, in cuml.internals.base.Base.__getattr__
AttributeError: coef_. Did you mean: '_coef_'?
```

I have stared at this for quite a while but can't work out what is going on??

Reply:
That is the same behavior a scikit-learn, no?

Reply:
I don't think so. I was trying to access the `coef_` attribute within `fit`.

In scikit-learn these are normal attributes, so once they are set you can use them. Right now we define `__getattribute__` which uses `_is_fit`. I think it is a bit weird to have code like this fail, mostly because it makes you question your sanity and because the exception doesn't contain a clue (we get to see the `AttributeError` from `__getattr__` not `__getattribute__` :():
```python
self.foo_ = 42
print(self.foo_) # Error, `foo_` doesn't exist!
```

Maybe we can get around the need to checking `_is_fit` and using `__getattribute__` by recording inside the `DynamicDescriptor` if it has been set or not:
```python
class DynamicDescriptor:
  def __init__(self, attribute_name):
    self.set = False
    self.attribute_name = f"{name}"

  def __get__(self, obj, objtype=None):
    if not self.set:
      raise AttributeError(f"{obj.__class__.__name__} object has no attribute {self.attribute_name}")
    else:
      return self.raw

  def __set__(self, obj, value):
    self.set = True
    # we can even store the value inside the descriptor?!
    self.raw = value
```

This might need a bit of tweaking to make the message in the exception look right (`"'Estimator' object has no attribute 'foo_'"`).